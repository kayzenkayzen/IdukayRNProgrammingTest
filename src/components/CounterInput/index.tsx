import React, { useCallback } from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { theme } from '@src/theme';
import { ICounterInput } from './index.d';

const styles = StyleSheet.create({
  container: { flexDirection: 'row', justifyContent: 'center', height: 40 },
  button: {
    width: 40,
    backgroundColor: '#000',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 4,
  },
  buttonLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
  },
  leftButton: {
    borderBottomLeftRadius: 20,
    borderTopLeftRadius: 20,
  },
  rightButton: {
    borderBottomEndRadius: 20,
    borderTopEndRadius: 20,
  },
  value: {
    width: 60,
    backgroundColor: '#fff',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: '#000',
    borderRadius: 4,
    marginHorizontal: 5,
  },
  valueLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
    color: '#000',
  },
});

const CounterInput: React.FC<ICounterInput> = React.memo(props => {
  const {
    id,
    buttonBackgroundColor,
    buttonLabelColor,
    buttonBorderColor,
    value,
    min,
    max,
    onPressMinus,
    onPressMore,
  } = props;

  const handleOnPressMinus = useCallback(() => {
    typeof min === 'undefined' || (typeof min !== 'undefined' && value > min)
      ? onPressMinus(id)
      : null;
  }, [onPressMinus, id, min, value]);

  const handleOnPressMore = useCallback(() => {
    typeof max === 'undefined' || (typeof max !== 'undefined' && value < max)
      ? onPressMore(id)
      : null;
  }, [onPressMore, id, max, value]);

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID="minus-button"
        style={[
          styles.button,
          styles.leftButton,
          {
            backgroundColor: buttonBackgroundColor,
            borderColor: buttonBorderColor,
          },
        ]}
        activeOpacity={1}
        onPress={handleOnPressMinus}
      >
        <Text
          style={[
            styles.buttonLabel,
            {
              color:
                typeof min === 'undefined' ||
                (typeof min !== 'undefined' && value > min)
                  ? buttonLabelColor
                  : buttonLabelColor + '30',
            },
          ]}
        >
          -
        </Text>
      </TouchableOpacity>

      <View
        style={[
          styles.value,
          {
            borderColor: buttonBorderColor,
          },
        ]}
      >
        <Text testID="value-label" style={[styles.valueLabel]}>
          {value}
        </Text>
      </View>

      <TouchableOpacity
        testID="more-button"
        style={[
          styles.button,
          styles.rightButton,
          {
            backgroundColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBackgroundColor
                : buttonBackgroundColor + '40',
            borderColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBorderColor
                : buttonBackgroundColor + '40',
          },
        ]}
        activeOpacity={0.8}
        onPress={handleOnPressMore}
      >
        <Text style={[styles.buttonLabel, { color: buttonLabelColor }]}>+</Text>
      </TouchableOpacity>
    </View>
  );
});

CounterInput.defaultProps = {
  buttonBackgroundColor: 'transparent',
  buttonBorderColor: '#000000',
  buttonLabelColor: '#000000',
};

export default CounterInput;
