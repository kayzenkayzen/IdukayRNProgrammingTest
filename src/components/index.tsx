export { default as CounterInput } from '@src/components/CounterInput';
