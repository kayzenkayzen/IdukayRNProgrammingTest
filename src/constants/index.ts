import { IAttackDamagePercentage, IPotion } from './index.d';

// Define the potions
const POTIONS: IPotion[] = [
  { id: 'red', rgb: '#BF2137', label: 'Poción Roja' },
  { id: 'blue', rgb: '#0479F1', label: 'Poción Azul' },
  { id: 'green', rgb: '#009E2F', label: 'Poción Verde' },
  { id: 'yellow', rgb: '#E4B80A', label: 'Poción Amarilla' },
  { id: 'gray', rgb: '#303030', label: 'Poción Gris' },
];

// Define the damage percentage
// for combinations length.
// The key is the length of the combination
// and the value is the percentage of damage
// for this combination length
const COMBINATION_DAMAGE_PERCENTAGE: IAttackDamagePercentage = {
  1: 3,
  2: 5,
  3: 10,
  4: 20,
  5: 25,
};

export const constants = {
  POTIONS,
  COMBINATION_DAMAGE_PERCENTAGE,
};
