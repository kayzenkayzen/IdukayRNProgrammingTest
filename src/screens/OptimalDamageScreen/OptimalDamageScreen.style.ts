import { StyleSheet } from 'react-native';
import { theme } from '@src/theme';

export const styles = StyleSheet.create({
  titleContainer: {
    paddingHorizontal: 10,
    paddingVertical: 10,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  titleText: {
    fontFamily: theme.fonts.primary.regular,
    fontSize: 16,
    lineHeight: 21,
    color: theme.colors.primary.medium,
    flex: 1,
  },
  damagePercentageOuterCircle: {
    width: 75,
    height: 75,
    borderRadius: 75,
    backgroundColor: 'green',
    justifyContent: 'center',
    alignItems: 'center',
    marginLeft: 20,
  },
  damagePercentageInnerCircle: {
    width: 70,
    height: 70,
    borderRadius: 70,
    backgroundColor: '#fff',
    justifyContent: 'center',
    alignItems: 'center',
  },
  damagePercentageCircleValue: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 21,
    textAlign: 'center',
    color: 'green',
  },
});
