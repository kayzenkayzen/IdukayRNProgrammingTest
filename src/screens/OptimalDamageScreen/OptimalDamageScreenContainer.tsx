import React from 'react';
import { OptimalDamageScreen } from '@src/screens';
import { IOptimalDamageScreenProps } from './types';

const OptimalDamageScreenContainer: React.FC<IOptimalDamageScreenProps> = React.memo(
  (props: IOptimalDamageScreenProps) => {
    return <OptimalDamageScreen {...props} />;
  },
);

export default OptimalDamageScreenContainer;
