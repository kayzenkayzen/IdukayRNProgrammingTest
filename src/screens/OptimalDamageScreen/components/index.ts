export { default as AttackCombinationItem } from './AttackCombinationItem/AttackCombinationItem';
