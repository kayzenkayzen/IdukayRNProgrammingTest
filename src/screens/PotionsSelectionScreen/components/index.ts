export { default as PotionOptionItem } from './PotionOptionItem/PotionOptionItem';
