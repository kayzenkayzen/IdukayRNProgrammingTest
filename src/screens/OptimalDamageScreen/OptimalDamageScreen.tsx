import React from 'react';
import { SafeAreaView, ScrollView, View, Text } from 'react-native';
import { theme } from '@src/theme';
import { AttackCombinationItem } from './components';
import { IOptimalDamageScreenProps } from './OptimalDamageScreen.d';
import { styles } from './OptimalDamageScreen.style';

const OptimalDamageScreen: React.FC<IOptimalDamageScreenProps> = React.memo(
  ({ route }: IOptimalDamageScreenProps) => {
    const { attacks, damage } = route.params;

    return (
      <SafeAreaView style={theme.container.outer}>
        <ScrollView>
          <View style={styles.titleContainer}>
            <Text style={styles.titleText}>
              Realiza estos ataques para lograr el mayor da√±o a tu objetivo
            </Text>

            <View style={styles.damagePercentageOuterCircle}>
              <View style={styles.damagePercentageInnerCircle}>
                <Text
                  testID="damage-value"
                  style={styles.damagePercentageCircleValue}
                >{`${damage} %`}</Text>
              </View>
            </View>
          </View>

          {(attacks || []).map((attack: string[], index: number) => {
            return (
              <AttackCombinationItem
                key={`attack-${index}`}
                attack={attack}
                index={index}
              />
            );
          })}
        </ScrollView>
      </SafeAreaView>
    );
  },
);

export default OptimalDamageScreen;
