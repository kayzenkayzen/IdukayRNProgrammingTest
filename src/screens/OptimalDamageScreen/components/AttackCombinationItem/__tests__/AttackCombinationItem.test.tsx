import React from 'react';
import { render } from '@testing-library/react-native';
import { constants } from '@src/constants';
import { AttackCombinationItem } from '@src/screens/OptimalDamageScreen/components';

describe('Test AttackCombinationItem Component', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <AttackCombinationItem
        attack={[constants.POTIONS[0].id, constants.POTIONS[1].id]}
        index={0}
      />,
    ).toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
