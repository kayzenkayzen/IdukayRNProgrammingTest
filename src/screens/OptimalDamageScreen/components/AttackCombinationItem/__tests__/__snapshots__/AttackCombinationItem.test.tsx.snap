// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AttackCombinationItem Component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderBottomColor": "#ddd",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingHorizontal": 10,
        "paddingVertical": 15,
      },
      Object {
        "borderTopColor": "#ddd",
        "borderTopWidth": 1,
      },
    ]
  }
  testID="attack-0"
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "marginRight": 10,
        "width": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "SourceSansPro-Bold",
          "fontSize": 16,
          "lineHeight": 26,
        }
      }
    >
      1
    </Text>
    <Text
      style={
        Object {
          "fontFamily": "SourceSansPro-Bold",
          "fontSize": 11,
          "lineHeight": 18,
        }
      }
    >
      o
    </Text>
  </View>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/potion.png",
      }
    }
    style={
      Array [
        Object {
          "height": 20,
          "marginRight": 10,
          "width": 20,
        },
        Object {
          "tintColor": "#BF2137",
        },
      ]
    }
    testID="attack-0-red-0"
  />
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/potion.png",
      }
    }
    style={
      Array [
        Object {
          "height": 20,
          "marginRight": 10,
          "width": 20,
        },
        Object {
          "tintColor": "#0479F1",
        },
      ]
    }
    testID="attack-0-blue-1"
  />
</View>
`;
